File: scip-java/src/main/scala/com/sourcegraph/scip_java/commands/SnapshotCommand.scala
package com.sourcegraph.scip_java.commands

import java.net.URI
import java.nio.charset.StandardCharsets
import java.nio.file._
import java.nio.file.attribute.BasicFileAttributes

import scala.jdk.CollectionConverters._

import com.sourcegraph.Scip
import com.sourcegraph.io.DeleteVisitor
import com.sourcegraph.scip_java.ScipPrinters
import moped.annotations._
import moped.cli.Application
import moped.cli.Command
import moped.cli.CommandParser

@Description(
  "Generates annotated snapshots for each `*.scip` file in the given target roots."
)
@Usage("scip-java snapshot [OPTIONS ...] [POSITIONAL ARGUMENTS ...]")
@ExampleUsage(
  "scip-java snapshot --output=generated/ my/targetroo1 my/targetroot2"
)
@CommandName("snapshot")
case class SnapshotCommand(
    @PositionalArguments
    @Description("List of directories containing SCIP files")
    targetroot: List[Path] = Nil,
    @Description("Output directory for the annotated snapshots")
    output: Path = Paths.get("generated"),
    cleanup: Boolean = true,
    @Inline()
    app: Application = Application.default
) extends Command {
  def sourceroot: Path = app.env.workingDirectory

  override def run(): Int = {
    val scipPattern = FileSystems.getDefault.getPathMatcher("glob:**.scip")
    if (cleanup) {
      Files.walkFileTree(output, new DeleteVisitor())
    }
    Files.createDirectories(output)
    var foundScipFile = false
    targetroot.foreach { root =>
      Files.walkFileTree(
        root,
        new SimpleFileVisitor[Path] {
          override def visitFile(
              file: Path,
              attrs: BasicFileAttributes
          ): FileVisitResult = {
            if (scipPattern.matches(file)) {
              val index = Scip.Index.parseFrom(Files.readAllBytes(file))
              // Skip per-source shards emitted by the compiler plugin (those don't have a
              // project_root). The aggregator produces a single top-level index file that
              // carries the project_root and is the canonical input for snapshot rendering.
              val rawProjectRoot = index.getMetadata.getProjectRoot
              if (rawProjectRoot.nonEmpty) {
                foundScipFile = true
                val projectRoot = URI.create(rawProjectRoot)
                index
                  .getDocumentsList
                  .asScala
                  .foreach { doc =>
                    val sourcepath = Paths.get(projectRoot.resolve(doc.getRelativePath))
                    val source =
                      new String(
                        Files.readAllBytes(sourcepath),
                        StandardCharsets.UTF_8
                      )
                    val document = ScipPrinters.printTextDocument(
                      doc,
                      source,
                      CommentSyntax.default
                    )
                    val snapshotOutput = output.resolve(doc.getRelativePath)
                    Files.createDirectories(snapshotOutput.getParent)
                    Files.write(
                      snapshotOutput,
                      document.getBytes(StandardCharsets.UTF_8)
                    )
                  }
              }
            }
            super.visitFile(file, attrs)
          }
        }
      )
    }
    if (foundScipFile) {
      0
    } else {
      app.error(
        s"no SCIP files found. To fix this problem, make sure that one of the directories " +
          s"in ${targetroot.mkString(", ")} contains a `*.scip` file."
      )
      1
    }
  }
}

object SnapshotCommand {
  implicit val parser = CommandParser.derive(SnapshotCommand())
}

File: scip-java/src/main/kotlin/com/sourcegraph/scip_java/commands/SnapshotCommand.kt
package com.sourcegraph.scip_java.commands

import com.github.ajalt.clikt.core.CliktCommand
import com.github.ajalt.clikt.core.Context
import com.github.ajalt.clikt.core.ProgramResult
import com.github.ajalt.clikt.core.requireObject
import com.github.ajalt.clikt.parameters.arguments.argument
import com.github.ajalt.clikt.parameters.arguments.multiple
import com.github.ajalt.clikt.parameters.options.default
import com.github.ajalt.clikt.parameters.options.flag
import com.github.ajalt.clikt.parameters.options.option
import com.github.ajalt.clikt.parameters.types.path
import com.sourcegraph.scip_java.ScipJavaApp
import com.sourcegraph.scip_java.ScipPrinters
import java.net.URI
import java.nio.charset.StandardCharsets
import java.nio.file.FileSystems
import java.nio.file.FileVisitResult
import java.nio.file.Files
import java.nio.file.Path
import java.nio.file.Paths
import java.nio.file.SimpleFileVisitor
import java.nio.file.attribute.BasicFileAttributes
import org.scip_code.scip.Index

/**
 * `scip-java snapshot`: generates annotated snapshots for each `*.scip`
 * file in the given target roots.
 */
class SnapshotCommand : CliktCommand(name = "snapshot") {

    override fun help(context: Context) =
        "Generates annotated snapshots for each `*.scip` file in the given target roots."

    private val app: ScipJavaApp by requireObject()

    private val targetroot: List<Path> by argument(
        help = "List of directories containing SCIP files",
    ).path().multiple()

    private val output: Path by option(
        "--output",
        help = "Output directory for the annotated snapshots",
    ).path().default(Paths.get("generated"))

    private val cleanup: Boolean by option(
        "--cleanup",
        "--no-cleanup",
    ).flag("--no-cleanup", default = true)

    override fun run() {
        val scipPattern = FileSystems.getDefault().getPathMatcher("glob:**.scip")
        if (cleanup) {
            output.toFile().deleteRecursively()
        }
        Files.createDirectories(output)
        var foundScipFile = false
        for (root in targetroot) {
            if (!Files.exists(root)) {
                app.warning("ignoring target root that does not exist: $root")
                continue
            }
            Files.walkFileTree(
                root,
                object : SimpleFileVisitor<Path>() {
                    override fun visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult {
                        if (scipPattern.matches(file)) {
                            val index = Index.parseFrom(Files.readAllBytes(file))
                            // Per-source SCIP shards under META-INF/scip/ carry no Metadata;
                            // only the aggregated index does. Skip shards so `scip-java
                            // snapshot <targetroot>` doesn't trip over them.
                            val projectRoot = index.metadata.projectRoot
                            if (projectRoot.isNotEmpty()) {
                                foundScipFile = true
                                val rootUri = URI.create(projectRoot)
                                for (doc in index.documentsList) {
                                    val sourcepath = Paths.get(rootUri.resolve(doc.relativePath))
                                    val source =
                                        String(Files.readAllBytes(sourcepath), StandardCharsets.UTF_8)
                                    val document = ScipPrinters.printTextDocument(doc, source)
                                    val snapshotOutput = output.resolve(doc.relativePath)
                                    Files.createDirectories(snapshotOutput.parent)
                                    Files.write(
                                        snapshotOutput,
                                        document.toByteArray(StandardCharsets.UTF_8),
                                    )
                                }
                            }
                        }
                        return super.visitFile(file, attrs)
                    }
                },
            )
        }
        if (!foundScipFile) {
            app.error(
                "no SCIP files found. To fix this problem, make sure that one of the directories " +
                    "in ${targetroot.joinToString(", ")} contains a `*.scip` file.",
            )
            throw ProgramResult(1)
        }
    }
}
